pkg/controller/util/endpoint/controller_utils.go

```diff
@@ -32,6 +34,99 @@ import (
 // EndpointsMatch is a type of function that returns true if pod endpoints match.
 type EndpointsMatch func(*v1.Pod, *v1.Pod) bool
 
+// ServiceSelectorCache is a cache of service selectors to avoid high CPU consumption caused by frequent calls to AsSelectorPreValidated (see #73527)
+type ServiceSelectorCache struct {
+	cache map[string]labels.Selector
+	lock  sync.RWMutex
+}
+
+// NewServiceSelectorCache init ServiceSelectorCache for both endpoint controller and endpointSlice controller.
+func NewServiceSelectorCache() *ServiceSelectorCache {
+	return &ServiceSelectorCache{
+		cache: map[string]labels.Selector{},
+	}
+}
+
+// Get return selector and existence in ServiceSelectorCache by key.
+func (sc *ServiceSelectorCache) Get(key string) (labels.Selector, bool) {
+	sc.lock.RLock()
+	defer sc.lock.RUnlock()
+	selector, ok := sc.cache[key]
+	return selector, ok
+}
+
+// Update can update selector of ServiceSelectorCache while service's selector changed.
+// Update can also add selector to ServiceSelectorCache while service's selector no cached.
+func (sc *ServiceSelectorCache) Update(key string, service *v1.Service) labels.Selector {
+	sc.lock.Lock()
+	defer sc.lock.Unlock()
+	v, ok := sc.cache[key]
+	// service selector not exist in ServiceSelectorCache, add it to ServiceSelectorCache.
+	if !ok {
+		selector := labels.Set(service.Spec.Selector).AsSelectorPreValidated()
+		sc.cache[key] = selector
+		return selector
+	}
+
+	// service selector exist in ServiceSelectorCache, update it in ServiceSelectorCache if selector changed.
+	oldSelector := v.(labels.Selector)
+	curSelector := labels.Set(service.Spec.Selector).AsSelectorPreValidated()
+	if reflect.DeepEqual(curSelector, oldSelector) {
+		return oldSelector
+	}
+	// service selector changed, so update ServiceSelectorCache
+	sc.cache[key] = curSelector
+	return curSelector
+}
+
+// Delete can delete selector which exist in ServiceSelectorCache.
+func (sc *ServiceSelectorCache) Delete(key string) {
+	sc.lock.Lock()
+	defer sc.lock.Unlock()
+	if _, ok := sc.cache[key]; !ok {
+		return
+	}
+	delete(sc.cache, key)
+	return
+}
+
+// GetPodServiceMemberships returns a set of Service keys for Services that have
+// a selector matching the given pod.
+func (sc *ServiceSelectorCache) GetPodServiceMemberships(serviceLister v1listers.ServiceLister, pod *v1.Pod) (sets.String, error) {
+	set := sets.String{}
+	services, err := serviceLister.Services(pod.Namespace).List(labels.Everything())
+	if err != nil {
+		// don't log this error because this function makes pointless
+		// errors when no services match
+		return set, nil
+	}
+
+	var selector labels.Selector
+	for _, service := range services {
+		key, err := controller.KeyFunc(service)
+		if err != nil {
+			return nil, err
+		}
+		if v, ok := sc.Get(key); ok {
+			selector = v
+		} else {
+			if service.Spec.Selector == nil {
```

Comment:
This seems like the very first thing to check at the top of this loop, since nothing else is relevant if the selector is nil (as long as I'm understanding this right).